### ISSUE TITLE
[no-unnecessary-condition] Option to ignore "both sides of the expression are literal" in ternary (?:) chains of enum cases

### ISSUE THREAD

**Issue by Jessidhia** [typescript-eslint/typescript-eslint#3637]
- [x] I have tried restarting my IDE and the issue persists.
- [ ] I have updated to the latest version of the packages.
- [x] I have [read the FAQ](https://github.com/typescript-eslint/typescript-eslint/blob/master/docs/getting-started/linting/FAQ.md) and my problem is not listed.

**Repro**

```JSON
{
  "rules": {
    "@typescript-eslint/no-unnecessary-condition": "error"
  }
}
```

```TS
enum Privacy {
  Public,
  Private
}

const stringify = (p: Privacy) => p === Privacy.Public ? 'Public' : p === Privacy.Private ? 'Private' : ''
```

**Expected Result**

Some way to allow `p === Privacy.Private` to not itself flag an error, as it's being used in an apparently redundant way as a way to ensure exhaustiveness. A `switch` is not necessarily available to use as `switch` is not an expression, while a `?:` chain is.

Here's an example in real code with a `JSXExpressionContainer` where the substitution cannot be made without either extracting the code into a one-off function call / component with a `switch` body, or using mutable `let` variables with mutation in a `switch` body done in a separate part of the component's render body:

![image](https://user-images.githubusercontent.com/73085/125728227-2b988270-86b4-4e7f-9c78-800fe9f8bff7.png)

**Actual Result**

"Unnecessary conditional, both sides of the expression are literal values" on the `p === Privacy.Private` check.

**Additional Info**

This is a bit of a rehash of https://github.com/typescript-eslint/typescript-eslint/issues/2952. I know that issue was closed, but the conversation in that issue was about choosing between `if` and `switch`, which are both `Statement`s; but `?:` is a `ConditionalExpression` and `Expression`s cannot be replaced by `Statement` outside of `Statement` contexts, and one such context where a `Statement` cannot be used is `JSXExpressionContainer`.

Not having an option to relax this, either heuristically for `enum === enum in ConditionalExpression` or for all `ConditionalExpression`s, would force the entire rule to be disabled due to how widespread this pattern is in our code, which loses all the benefits the rule brings to `if` statements and other conditional checks that don't involve `enum`s...

**Versions**

| package                            | version |
| ---------------------------------- | ------- |
| `@typescript-eslint/eslint-plugin` | `4.13.0` |
| `@typescript-eslint/parser`        | `4.13.0` |
| `TypeScript`                       | `4.1.2` |
| `ESLint`                           | `7.15.0` |
| `node`                             | `14.15.4` |

**Comment by bradzacher** [comment]
There's no way to specifically target the case of "this thing has already been narrowed down to just compare this one case". All we can see is that the check is unnecessary.

Which means that is literally the only thing we could do with the option. "Are both the left and right side an enum and the check is unnecessary?"

As I mentioned in my closing comment on the issue - it's not a good idea to do this because of the false-negative rate.
In your examples for sure you've got all your checks inline in a single ternary. But what happens if your code does separate checks, should that be allowed?

```ts
enum Mode { one, two }

function test(condition: boolean, mode: Mode) {
  if (mode === Mode.one) {
    return;
  }

  if (condition) {
  } else if (mode === Mode.two) {} // should this be ignored by the rule??

  if (condition || mode === Mode.two) {} // should this be ignored by the rule??
}
```

To me those would be clear false negatives, same as if they were instead this:
```ts
function test(condition: boolean, mode: string | null) {
  if (typeof mode === 'string') {
    return;
  }

  if (condition) {
  } else if (mode === null) {} // clearly unnecessary

  if (condition || mode === null) {} // clearly unnecessary
}
```

And false-negatives harm user trust in linting. False negatives are insidious because you don't know they're there until you figure out that they're there. This makes you question the lint rule or even the entire plugin. "If this rule missed this case, what other cases has the linter missed?"

Ultimately I'm still against this kind of option.
I don't think weakening the rule with this option leads to anything good within a codebase.
I think it leads just to allowing ambiguous code alongside bad coding patterns.

We purposely licence this project very permissively so that others can take the code and fork it as appropriate. If this is a pattern you really want to use in your project, then it might be a good idea to fork the rule locally and change it to suit your project's code style.
As an example we do this a lot internally at Facebook because there are many good lint rules that don't quite fit the style in the company's huge codebase instead of skipping a rule, we make it work instead.

----

The following is tangential to the issue, and is more in response to your Additional Info paragraph.

Code is written once and read many times, and so more often than not terse code is bad code as it actively harms readability.

IMO Nested ternaries are the enemy of readability, so reaching for them - ESP when solving for unrelated logic like your second example does - just makes your code harder to understand.

The question I ask here is "why do you need to specifically use an expression?"

Your first example could have easily been an arrow function with a body using a switch. There's no reason it has to be a body-less arrow function aside from coding tersely.

Your second example ofc is in a JSX expression container, meaning it does have to be an expression.

Though there's likely no explicit reason that this logic has to be inlined aside from code colocation - i.e. it could be extracted out to a variable, and assigned via statements.

If it has to be inlined due to some other logic, then an IIFE is what I reach for to allow me to use better code patterns and write clearer code:
```ts
(() => {
  switch (status) {
    case 'loading':
      return <LoadingContainer />;

    case 'notFound':
      return <WorksNotFound />;
  }

  switch (props.workType) {
    case WorkType.Illust:
      return <GridIllust ... />;

    case WorkType.Novel:
      return <GridNovel ... />;

    // ...
  }
})()
```

At some point in the distant future you might instead be able to reach for a [do expression](https://github.com/tc39/proposal-do-expressions), but for now IMO an IIFE is a great tool that helps write clearer code.

**Comment by bradzacher** [comment]
@ypresto - that seems very much like a bug in TypeScript, and not something for us to cover with lint rules.

Based on my comment above - we will not be adding this option.
Thanks for the suggestion!

**Comment by ypresto** [comment]
@bradzacher It seems not a bug, but is design.

TypeScript enum is used as bit flags, so its value will go outside of predefined integer values.
https://basarat.gitbook.io/typescript/type-system/enums#number-enums-as-flags

Issue for this in TypeScript repo is marked as a "suggestion", not a bug.
https://github.com/microsoft/TypeScript/issues/32690
